api/v1alpha1/extension_types.go

```diff
@@ -0,0 +1,133 @@
+	ManagedStateActive ExtensionManagedState = "Active"
+	// Peform resolution of this Extension
+	ManagedStatePaused ExtensionManagedState = "Paused"
+)
+
+type ExtensionSourcePackage struct {
+	//+kubebuilder:validation:MaxLength:=48
+	//+kubebuilder:validation:Pattern:=^[a-z0-9]+(-[a-z0-9]+)*$
+	Name string `json:"name"`
+
+	//+kubebuilder:validation:MaxLength:=64
+	//+kubebuilder:validation:Pattern=`^(\s*(=||!=|>|<|>=|=>|<=|=<|~|~>|\^)\s*(v?(0|[1-9]\d*|[x|X|\*])(\.(0|[1-9]\d*|x|X|\*]))?(\.(0|[1-9]\d*|x|X|\*))?(-([0-9A-Za-z\-]+(\.[0-9A-Za-z\-]+)*))?(\+([0-9A-Za-z\-]+(\.[0-9A-Za-z\-]+)*))?)\s*)((?:\s+|,\s*|\s*\|\|\s*)(=||!=|>|<|>=|=>|<=|=<|~|~>|\^)\s*(v?(0|[1-9]\d*|x|X|\*])(\.(0|[1-9]\d*|x|X|\*))?(\.(0|[1-9]\d*|x|X|\*]))?(-([0-9A-Za-z\-]+(\.[0-9A-Za-z\-]+)*))?(\+([0-9A-Za-z\-]+(\.[0-9A-Za-z\-]+)*))?)\s*)*$`
+	//+kubebuilder:Optional
+	// Version is an optional semver constraint on the package version. If not specified, the latest version available of the package will be installed.
+	// If specified, the specific version of the package will be installed so long as it is available in any of the content sources available.
+	// Examples: 1.2.3, 1.0.0-alpha, 1.0.0-rc.1
+	//
+	// For more information on semver, please see https://semver.org/
+	Version string `json:"version,omitempty"`
+
+	//+kubebuilder:validation:MaxLength:=48
+	//+kubebuilder:validation:Pattern:=^[a-z0-9]+([\.-][a-z0-9]+)*$
+	// Channel constraint definition
+	Channel string `json:"channel,omitempty"`
+}
+
+// TODO: Implement ExtensionSourceDirect containing a URL or other reference mechanism
+
+type ExtensionSource struct {
+	// A source package defined by a name, version and/or channel
+	Package *ExtensionSourcePackage `json:"package,omitempty"`
+
+	// A direct package defined by a URL
+	// TODO: Direct *ExtensionSourceDirect `json:"direct,omitempty"`
+}
+
+// ExtensionSpec defines the desired state of Extension
+type ExtensionSpec struct {
+	//+kubebuilder:validation:Enum:=Active;Paused
+	//+kubebuilder:default:=Active
+	//+kubebuilder:Optional
+	//
+	// Pause reconciliation on this Extension
+	Managed ExtensionManagedState `json:"managed,omitempty"`
+
+	//+kubebuilder:validation:MaxLength:=64
+	//+kubebuilder:validation:Pattern:=^[a-z0-9]+([\.-][a-z0-9]+)*$
+	//
+	// ServiceAccount name used to install this extension
+	ServiceAccountName string `json:"serviceAccountName"`
+
+	//+kubebuilder:validation:MaxLength:=64
+	//+kubebuilder:validation:Pattern:=^[a-z0-9]+([\.-][a-z0-9]+)*$
+	//+kubebuilder:Optional
+	//
+	// Location of installation TBD??
+	DefaultNamespace string `json:"defaultNamespace,omitempty"`
+
+	// Source of Extension to be installed
+	Source ExtensionSource `json:"source"`
+
+	//+kubebuilder:validation:Enum:=Enforce;Ignore
+	//+kubebuilder:default:=Enforce
+	//+kubebuilder:Optional
+	//
+	// Defines the policy for how to handle upgrade constraints
+	UpgradeConstraintPolicy UpgradeConstraintPolicy `json:"upgradeConstraintPolicy,omitempty"`
```

Comment:
What is our perception of when this field will be applicable?

If we add a way to directly reference a bundle in the future, that means we are not using a catalog. However a catalog is where we (currently) encode all upgrade edge information. There is no upgrade edge information on individual bundles, and there is no implicit versioning scheme (like semver) today. Ergo, there is no way to implement upgrade constraint policies without a catalog.

Should we move this field into the `ExtensionSourcePackage` struct?